HW2/P3/mandelbrot.pyx

```diff
@@ -0,0 +1,148 @@
+       #     v3 = AVX.mul(v1, v2)
+       #
+       # You may find the numpy.real() and numpy.imag() fuctions helpful.
+
+    assert in_coords.shape[1] % 8 == 0, "Input array must have 8N columns"
+    assert in_coords.shape[0] == out_counts.shape[0], "Input and output arrays must be the same size"
+    assert in_coords.shape[1] == out_counts.shape[1],  "Input and output arrays must be the same size"
+
+    with nogil:
+       
+        for i in prange(in_coords.shape[0], schedule='static',chunksize=1,num_threads = 4):
+            for j in range(in_coords.shape[1]/8):
+                
+                iterations = AVX.float_to_float8(0.0)
+                ones = AVX.float_to_float8(1.0)
+                
+                #We create two arrays of 8 elements, one for the real part of the complex numbers and one for the imaginary part.
+                c_real = AVX.make_float8((in_coords[i,j*8+7]).real,
+                             (in_coords[i,j*8+6]).real,
+                             (in_coords[i,j*8+5]).real,
+                             (in_coords[i,j*8+4]).real,
+                             (in_coords[i,j*8+3]).real,
+                             (in_coords[i,j*8+2]).real,
+                             (in_coords[i,j*8+1]).real,
+                             (in_coords[i,j*8]).real)
+                
+                c_imag = AVX.make_float8((in_coords[i,j*8+7]).imag,
+                             (in_coords[i,j*8+6]).imag,
+                             (in_coords[i,j*8+5]).imag,
+                             (in_coords[i,j*8+4]).imag,
+                             (in_coords[i,j*8+3]).imag,
+                             (in_coords[i,j*8+2]).imag,
+                             (in_coords[i,j*8+1]).imag,
+                             (in_coords[i,j*8]).imag)
+                
+               
+                z_real = AVX.float_to_float8(0.0)
+                z_imag = AVX.float_to_float8(0.0)
+                
+                
+                
+                for iter in range(max_iterations):
+                    
+                    #We flag the elements in the array in which magnitude_squared(element) > 4:
+                    mask = AVX.greater_than(AVX.float_to_float8(4), magnitude_squared(z_real, z_imag))
+                    
+                    #We invert mask and select off values. The array has ones for the elements in which magnitude_squared(element) <= 4 and 0 for the rest.
+                    accepted = AVX.bitwise_and(mask, ones)
+                    
+                    #We update in each iteration the values of accepted so we keep track of the numbers of iterations an element has had magnitude_squared(element) <= 4. 
+                    iterations = AVX.add(accepted, iterations)
+                        
+                    #If all the elements have values such that magnitude_squared(element) > 4 we stop iterating. 
+                    if AVX.signs(mask) == 0:
+                        break
+                    
```

Comment:
Nice, once again, I implemented the same break statement. I found that this helped the performance of the mandelbrot as there are no wasted calculations.